File: php/testcode/headerinj_004.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_headerinj_urlencode
function headerinj004(BenchmarkRequest $req): BenchmarkResponse {
    $username = $req->param('username');
    header("X-User: " . urlencode($username)); // vuln-code-snippet safe-line php_headerinj_urlencode
    return BenchmarkResponse::ok('Header set');
}
// vuln-code-snippet end php_headerinj_urlencode

File: ruby/testcode/headerinj_004.rb
require_relative 'shared'

ALLOWED_THEMES = %w[light dark high-contrast].freeze

# vuln-code-snippet start ruby_headerinj_allowlist_header
def set_theme_header(req)
  requested = req.param('theme')
  theme = ALLOWED_THEMES.include?(requested) ? requested : 'light'
  headers = { 'X-Theme' => theme }  # vuln-code-snippet safe-line ruby_headerinj_allowlist_header
  BenchmarkResponse.new(200, 'ok', headers)
end
# vuln-code-snippet end ruby_headerinj_allowlist_header
